File: sdk/go/result.go
package shojiku

// Result is what every lifecycle operation returns beside its error.
//
// The error and the result carry different kinds of bad news, and keeping
// them apart is the whole point of this type. A returned error means the
// CALLER got something wrong, or the transport got no answer. A Result whose
// [Result.Success] is false means the DOCUMENT was refused — a template that
// will not render, a key that will not sign, a signature that does not
// verify. Those are data you query, because a caller who checked only err
// would otherwise be told a forgery is fine.
//
// Diagnostics ride on a SUCCESS too. A render that worked can still have
// warned about an overflowing box, and a caller that only looks at failures
// never sees them.
type Result struct {
	artifact    *DocumentArtifact
	report      *VerificationReport
	diagnostics []Diagnostic
	failure     *Failure
}

func succeededWithArtifact(artifact *DocumentArtifact, diagnostics []Diagnostic) *Result {
	return &Result{artifact: artifact, diagnostics: diagnostics}
}

func succeededWithReport(report *VerificationReport, diagnostics []Diagnostic) *Result {
	return &Result{report: report, diagnostics: diagnostics}
}

func fromFailure(failure *Failure) *Result {
	return &Result{diagnostics: failure.diagnostics, failure: failure}
}

// refused is a verdict that FAILED but still carries its report — the whole
// reason NotChecked exists. Nothing else produces a failed result with a
// value.
func refused(report *VerificationReport, diagnostics []Diagnostic, failure *Failure) *Result {
	return &Result{report: report, diagnostics: diagnostics, failure: failure}
}

// Success reports whether the operation produced what was asked for.
func (r *Result) Success() bool { return r.failure == nil }

// Failed reports whether the document was refused.
func (r *Result) Failed() bool { return r.failure != nil }

// Artifact is the rendered or signed document, or nil when this result is
// not about one. It does not fail; [Result.Err] is the one that does.
func (r *Result) Artifact() *DocumentArtifact { return r.artifact }

// Report is the verification report, or nil when this result is not about
// one. A FAILING verdict still carries it, because what was not checked has
// to reach the caller either way.
func (r *Result) Report() *VerificationReport { return r.report }

// Err is nil on success and the failure as an error on failure.
//
// This is Go's form of the opt-in unwrap every Shojiku SDK offers: an
// accessor that surfaces the failure for a script that wants one line of
// control flow rather than a branch. The other six raise here; the ruling for
// Go is frozen as an error return rather than a panic, because the language
// has no exceptions to mirror. The name is Err rather than Unwrap because
// errors.Unwrap owns that verb in Go, while Err is the language's own
// spelling for "the error this value carries" (sql.Rows.Err, bufio.Scanner.Err).
//
// Application code that handles failure keeps using [Result.Success] and
// [Result.Failure]; nothing in this package calls this.
func (r *Result) Err() error {
	if r.failure == nil {
		return nil
	}
	return &UnwrapError{Failure: r.failure}
}

// Failure is why the operation did not produce what was asked for, or nil.
func (r *Result) Failure() *Failure { return r.failure }

// Diagnostics are everything the engine noticed, on a success as well as a
// failure.
func (r *Result) Diagnostics() []Diagnostic { return r.diagnostics }

// Errors are only the diagnostics that are errors — the ones that explain a
// refusal.
func (r *Result) Errors() []Diagnostic { return r.filter(Diagnostic.IsError) }

// Warnings are only the warnings, which a SUCCESSFUL result can carry.
func (r *Result) Warnings() []Diagnostic { return r.filter(Diagnostic.IsWarning) }

func (r *Result) filter(keep func(Diagnostic) bool) []Diagnostic {
	var kept []Diagnostic
	for _, d := range r.diagnostics {
		if keep(d) {
			kept = append(kept, d)
		}
	}
	return kept
}

File: sdk/php/src/Result.php
<?php

declare(strict_types=1);

namespace Shojiku;

use Shojiku\Exception\UnwrapException;

/**
 * What every lifecycle operation returns.
 *
 * Nothing in the normal flow throws. A template that will not render, a key
 * that will not sign, a signature that does not verify are all data you query
 * — `success()`, the value, the engine's diagnostics either way, and on
 * failure the {@see Failure} trace.
 *
 * Diagnostics ride on a SUCCESS too. A render that worked can still have
 * warned about an overflowing box, and a caller that only looks at failures
 * never sees them.
 */
final class Result
{
    /**
     * @param list<Diagnostic> $diagnostics
     */
    private function __construct(
        private readonly DocumentArtifact|VerificationReport|null $value,
        private readonly array $diagnostics,
        private readonly ?Failure $failure,
    ) {
    }

    /**
     * @param list<Diagnostic> $diagnostics
     */
    public static function succeeded(
        DocumentArtifact|VerificationReport|null $value,
        array $diagnostics,
    ): self {
        return new self($value, $diagnostics, null);
    }

    /**
     * `fromFailure` rather than the reference's `failed`, because a static
     * constructor and an instance predicate share one namespace in PHP and
     * the PREDICATE is the one that carries the contract's name. Python
     * settled the same collision the same way, and this mirrors its spelling
     * rather than inventing a third.
     */
    public static function fromFailure(Failure $failure): self
    {
        return new self(null, $failure->diagnostics(), $failure);
    }

    /**
     * A verdict that FAILED but still carries its report — the whole reason
     * `notChecked` exists. Nothing else produces a failed result with a
     * value.
     *
     * @param list<Diagnostic> $diagnostics
     */
    public static function refused(
        ?VerificationReport $report,
        array $diagnostics,
        Failure $failure,
    ): self {
        return new self($report, $diagnostics, $failure);
    }

    public function success(): bool
    {
        return $this->failure === null;
    }

    public function failed(): bool
    {
        return $this->failure !== null;
    }

    public function value(): DocumentArtifact|VerificationReport|null
    {
        return $this->value;
    }

    /**
     * The value under the name of what the operation produced. The same
     * object; the aliases exist so calling code reads as what it is doing,
     * and neither throws — {@see self::unwrap()} is the one that does.
     */
    public function artifact(): ?DocumentArtifact
    {
        return $this->value instanceof DocumentArtifact ? $this->value : null;
    }

    public function report(): ?VerificationReport
    {
        return $this->value instanceof VerificationReport ? $this->value : null;
    }

    /**
     * The value, or a thrown {@see UnwrapException} when the operation failed.
     *
     * The opt-in bridge for a script that wants a stack trace rather than a
     * branch, and the ONE place this API throws for something other than a
     * misused argument. The ruling is frozen for every Shojiku SDK: calling
     * unwrap on a failed result is programmer misuse — a caller who has not
     * checked `success()` is asserting the operation worked. Application code
     * that handles failure keeps using `success()` and `failure()`; nothing
     * in this package calls this.
     *
     * @throws UnwrapException
     */
    public function unwrap(): DocumentArtifact|VerificationReport|null
    {
        if ($this->failure !== null) {
            throw new UnwrapException($this->failure);
        }

        return $this->value;
    }

    public function failure(): ?Failure
    {
        return $this->failure;
    }

    /**
     * @return list<Diagnostic>
     */
    public function diagnostics(): array
    {
        return $this->diagnostics;
    }

    /**
     * Only the diagnostics that are errors — the ones that explain a refusal.
     *
     * @return list<Diagnostic>
     */
    public function errors(): array
    {
        return array_values(array_filter($this->diagnostics, static fn (Diagnostic $d) => $d->isError()));
    }

    /**
     * Only the warnings, which a SUCCESSFUL result can carry.
     *
     * @return list<Diagnostic>
     */
    public function warnings(): array
    {
        return array_values(array_filter($this->diagnostics, static fn (Diagnostic $d) => $d->isWarning()));
    }
}
